src/Platforms/MySQL/CharsetMetadataProvider/ConnectionCharsetMetadataProvider.php

```diff
@@ -0,0 +1,44 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Doctrine\DBAL\Platforms\MySQL\CharsetMetadataProvider;
+
+use Doctrine\DBAL\Connection;
+use Doctrine\DBAL\Exception;
+use Doctrine\DBAL\Platforms\MySQL\CharsetMetadataProvider;
+
+/**
+ * @internal
+ */
+final class ConnectionCharsetMetadataProvider implements CharsetMetadataProvider
+{
+    private Connection $connection;
+
+    public function __construct(Connection $connection)
+    {
+        $this->connection = $connection;
+    }
+
+    /**
+     * @throws Exception
+     */
+    public function getDefaultCharsetCollation(string $charset): ?string
+    {
+        $collation = $this->connection->fetchOne(
+            <<<'SQL'
+SELECT DEFAULT_COLLATE_NAME
+FROM information_schema.CHARACTER_SETS
+WHERE CHARACTER_SET_NAME = ?;
+SQL
```

Comment:
You are allowed to indent HEREDOC/NOWDOC blocks since PHP 7.3.

Reply:
I know. See https://github.com/doctrine/dbal/pull/4548#discussion_r596101728. If you both suggest using this syntax, we need to enforce it at the coding standard level.